rust/lance/src/dataset.rs

```diff
@@ -1460,6 +1458,121 @@ impl Dataset {
     }
 }
 
+pub(crate) struct NewTransactionResult<'a> {
+    pub dataset: BoxFuture<'a, Result<Dataset>>,
+    pub new_transactions: BoxStream<'a, Result<(u64, Arc<Transaction>)>>,
+}
+
+pub(crate) fn load_new_transactions(dataset: &Dataset) -> NewTransactionResult<'_> {
+    // Re-use the same list call for getting the latest manifest and the metadata
+    // for all manifests in between.
+    let io_parallelism = dataset.object_store().io_parallelism();
+    let latest_version = dataset.manifest.version;
+    let locations = dataset
+        .commit_handler
+        .list_manifest_locations(&dataset.base, dataset.object_store(), true)
+        .try_take_while(move |location| {
+            futures::future::ready(Ok(location.version > latest_version))
```

Comment:
maybe in future iterations, technically we can also push this down so that the list operation can just list manifests that are after the `latest_version` and avoid listing all manifests in the version directory.

Reply:
The return value of `list_manifest_locations` is a stream, so it should be efficient when it can be (list operation has lexical order and we are using V2 manifest paths). But often it will be unavoidable to list all manifests.

Though thinking about it, it might be worthwhile doing a fast path at least for local filesystem. The existing implementation finds the latest version (which has a fast local implementation) and then opens all the versions in between. That's probably what we should do for local.

Reply:
Would it be possible to leverage the start-after header in S3 when possible so we don't need to list 
anything before the latest version?
 https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_RequestSyntax